File: problems/13.priority-queue/ts/PriorityQueue.ts
export interface PriorityQueue {
    insert: (value: number) => void;
    getHeap: () => number[];
    getSize: () => number;
    peek: () => number;
    remove: (index: number) => void;
}

export class PriorityQueueImpl implements PriorityQueue {
    constructor(
        private heap: number[],
        private competitor: (a: number, b: number) => boolean
    ) {
        this.buildHeap();
    }

    public insert(value: number) {
        this.heap.push(value);
        this.heapifyUp(this.heap.length - 1);
    }

    public remove(index: number) {
        this.heap = this.heap.filter(
            (item: number, idx: number) => idx !== index
        );

        this.heapifyDown(index);
    }

    public getHeap() {
        return this.heap;
    }

    public getSize() {
        return this.heap.length;
    }

    public peek() {
        if (this.heap.length === 0) {
            throw new Error('Heap is empty');
        }

        const root = this.heap[0];
        this.heap[0] = this.heap.pop() as number;
        this.heapifyDown(0);

        return root;
    }

    private buildHeap() {
        for (let i = Math.floor((this.heap.length - 1) / 2); i >= 0; --i) {
            this.heapifyDown(i);
        }
    }

    private heapifyUp(index: number) {
        if (index === 0) return;

        const parentIndex = Math.floor((index - 1) / 2);

        if (!this.competitor(this.heap[parentIndex], this.heap[index])) {
            let temp = this.heap[index];
            this.heap[index] = this.heap[parentIndex];
            this.heap[parentIndex] = temp;
            this.heapifyUp(parentIndex);
        }
    }

    private heapifyDown(index: number) {
        const leftIndex = index * 2 + 1;
        const rightIndex = index * 2 + 2;
        const left = this.heap[leftIndex];
        const right = this.heap[rightIndex];

        // min or max
        const targetChildIndex = this.competitor(left, right)
            ? leftIndex
            : rightIndex;

        if (this.competitor(this.heap[targetChildIndex], this.heap[index])) {
            let temp = this.heap[index];
            this.heap[index] = this.heap[targetChildIndex];
            this.heap[targetChildIndex] = temp;
            this.heapifyDown(targetChildIndex);
        }
    }
}

File: problems/13.priority-queue/golang/priority_queue.go
package main

import (
	"errors"
	"math"
)

type PriorityQueue struct {
	heap       []int
	competitor func(a, b int) bool
}

func NewPriorityQueue(initial []int, competitor func(a, b int) bool) *PriorityQueue {
	instance := PriorityQueue{
		heap:       []int{},
		competitor: competitor,
	}

	instance.buildHeap(initial)
	return &instance
}

func (p *PriorityQueue) insert(value int) *PriorityQueue {
	p.heap = append(p.heap, value)
	p.heapifyUp(len(p.heap) - 1)
	return p
}

func (p *PriorityQueue) buildHeap(initial []int) {
	for _, value := range initial {
		p.heap = append(p.heap, value)
		p.heapifyUp(len(p.heap) - 1)
	}
}

func (p *PriorityQueue) heapifyUp(idx int) {
	if idx == 0 || p.isEmpty() {
		return
	}
	parentIndex := parent(idx)

	//if p.heap[parentIndex] > p.heap[idx] {
	if p.competitor(p.heap[parentIndex], p.heap[idx]) {
		temp := p.heap[idx]
		p.heap[idx] = p.heap[parentIndex]
		p.heap[parentIndex] = temp
		p.heapifyUp(parentIndex)
	}
}

func leftChild(idx int) int {
	return idx*2 + 1
}

func rightChild(idx int) int {
	return idx*2 + 2
}

func (p *PriorityQueue) heapifyDown(idx int) {
	leftChildIndex := leftChild(idx)
	rightChildIndex := rightChild(idx)
	target := rightChildIndex

	if p.heap[leftChildIndex] < p.heap[rightChildIndex] {
		target = leftChildIndex
	}

	if p.competitor(p.heap[idx], p.heap[target]) {
		temp := p.heap[idx]
		p.heap[idx] = p.heap[target]
		p.heap[target] = temp
		p.heapifyDown(target)
	}
}

func parent(index int) int {
	return int(math.Floor((float64(index-1) / 2)))
}

func (p *PriorityQueue) getHeap() []int {
	return p.heap
}

func (p *PriorityQueue) isEmpty() bool {
	return len(p.heap) == 0
}

func (p *PriorityQueue) getSize() int {
	return len(p.heap)
}

func (p *PriorityQueue) peek() (int, error) {
	if p.isEmpty() {
		return 0.0, errors.New("Empty heap")
	}

	root := p.heap[0]
	poppedValue := p.heap[len(p.heap)-1]
	p.heap = p.heap[:len(p.heap)-1]
	p.heap[0] = poppedValue
	p.heapifyDown(0)

	return root, nil
}

func (p *PriorityQueue) remove(idx int) {
	p.heap[idx] = p.heap[len(p.heap)-1]
	p.heap = p.heap[:len(p.heap)-1]
	p.heapifyDown(idx)
}

func (p *PriorityQueue) pop() (int, error) {
	if p.isEmpty() {
		return 0.0, errors.New("Empty heap")
	}
	removed := p.heap[len(p.heap)-1]
	p.remove(len(p.heap) - 1)
	return removed, nil
}
